dep0/lib/05_llvmgen/src/gen.cpp

```diff
@@ -790,21 +989,45 @@ llvm::CallInst* gen_func_call(
                     [] (typecheck::kind_t) { return false; });
             })
         and "can only generate function call for 1st order applications");
-    auto const call =
-        builder.CreateCall(
-            callee.type,
-            callee.func,
-            fmap(args, [&] (typecheck::expr_t const& arg)
-            {
-                return gen_val(global, local, builder, arg);
-            }));
-    for (auto const i: std::views::iota(0ul, args.size()))
+    // might need to emit an alloca
+    auto const func_type = std::get_if<typecheck::expr_t>(&app.func.get().properties.sort.get());
+    assert(func_type and "functions must be of sort type");
+    auto const pi_type = std::get_if<typecheck::expr_t::pi_t>(&func_type->value);
+    assert(pi_type and "functions must be pi-types");
+    auto const dest = gen_alloca_if_needed(global, local, builder, pi_type->ret_type.get());
+    auto const func = [&]
     {
-        auto const& arg_type = std::get<typecheck::expr_t>(args[i].properties.sort.get());
+        if (auto const abs = std::get_if<typecheck::expr_t::abs_t>(&app.func.get().value))
+        {
+            auto proto = llvm_func_proto_t::from_abs(*abs);
+            assert(proto and "can only invoke a 1st order function type");
+            return gen_func(global, local, *proto, *abs);
+        }
```

Comment:
should be able to move this inside `gen_val()` now

Reply:
A test like this should work, as normalisation will synthethise a lambda abstraction:
```
auto select(book which) { if (which) return f; else return g; }
auto choose(bool X) -> array_t((i32_t)->i32_t, 2)
{
    return {select(X)), select(negate(X))};
}
```